### PR TITLE
feat(e2e): adding playwright and migrate landing page e2e tests

### PR DESCRIPTION
**Pull request by raisedadead**
A sanity check for #51129, all this does is enables the workflow on a special `e2e-**` branch namespace.

### CODE REVIEW

**Comment by raisedadead** [comment]
Alright, I have ported the PR over to a branch on the main repo so that we can see the tests going. Without it, there wasn't a very good way to compare this with the real thing. 

But as soon as we are happy with the changes, I will FF this into the main branch.

File: .github/workflows/e2e-playwright.yml [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+          path: client-artifact.tar
+
+      - name: Upload Webpack Stats
+        uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # v3
+        with:
+          name: webpack-stats
+          path: client/public/stats.json
+
+  playwright-run:
+    name: Test
```
Comment (raisedadead) [inline .github/workflows/e2e-playwright.yml:74]:
```suggestion
    name: E2E Test
```

This job name conflicts with an existing job in the Node.js tests workflow making leading to a collision when we want to apply branch rules.

File: .github/workflows/e2e-playwright.yml [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    runs-on: ubuntu-20.04
+    needs: build-client
+    strategy:
+      matrix:
+        node-version: [18.x]
```
Comment (raisedadead) [inline .github/workflows/e2e-playwright.yml:79]: I don't see any strategy for different browsers. Can we run things in parallel jobs?

Comment (Ahmadkashif) [reply .github/workflows/e2e-playwright.yml:79]: Well that can be flaky, we could break this into chrome, webkit and firefox jobs. That tends to flaky results on local machine, haven't tried that out in github actions

File: .github/workflows/e2e-playwright.yml (hunk 1 above)
Comment (Ahmadkashif) [reply .github/workflows/e2e-playwright.yml:74]: Alright. Sounds good.

Comment (Ahmadkashif) [reply .github/workflows/e2e-playwright.yml:74]: do I make a new commit for it? @raisedadead

Comment (Ahmadkashif) [reply .github/workflows/e2e-playwright.yml:74]: @raisedadead here, added the suggestion

https://github.com/freeCodeCamp/freeCodeCamp/pull/51219

File: .github/workflows/e2e-playwright.yml (hunk 2 above)
Comment (raisedadead) [reply .github/workflows/e2e-playwright.yml:79]: Could you elaborate again, or point me to where I can learn more about it?

Comment (Ahmadkashif) [reply .github/workflows/e2e-playwright.yml:79]: Sure @raisedadead ,

I would like you to take a look at this conversation: https://github.com/microsoft/playwright/issues/15691

I shared this on discord as well if you recall?

Comment (Ahmadkashif) [reply .github/workflows/e2e-playwright.yml:79]: @raisedadead 

We can definitely figure out more ways to speed up the execution. I however think the first thing is to find merit in using playwright as a tool.

This is merely the starting point.

Comment (raisedadead) [reply .github/workflows/e2e-playwright.yml:79]: Thanks.

Let me rephrase my queries. 

I want to avoid reading random threads and issues and get lost in their contexts. So to correct my assumptions, please consider answering my specific questions:

- You are saying the tests are flaky locally. Can you **elaborate** on your setup? I assume you are running a single server against multiple clients (3 types of browsers). Is that correct?
- Can you point me to docs/faqs (not issue threads that do not make sense in our context) where Playwrite recommends not using parallel runs in CI?

While strategies are usually used to speed things up, the main reasons are we want to run specs in an isolated environment. By using a matrix we are able to do that across browsers. That way, we can pinpoint device-specific issues better.

Comment (Ahmadkashif) [reply .github/workflows/e2e-playwright.yml:79]: @raisedadead.

Playwright docs don't specify running parallel runs in CI. They claim it should work out of the box as written.

That same could be said about regular tests. While the documentaion claims they should work, I kept getting errors, on gitpod and locally, where I try and run tests in default mode ( parallel runs ).
 
My local setup is as follows: I run the `pnpm run develop` command and that serves the client at port 8000. I then run playwright tests. The current configuration lets me test all browsers via the CLI command.

Let me explain. The command npx playwright test will simply start running tests, one by one, for every browser within the same process. If, I allow parallel runs, the CLI runs multiple browsers at the same time against the same localhost server running on port 8000.

Comment (Ahmadkashif) [reply .github/workflows/e2e-playwright.yml:79]: @raisedadead 

However, I can improve the github action and try different things out.

I can try out what you are asking, to have isolated tests for chrome, firefox and webkit on the action. 

Let me know, if that is indeed what you are asking for? 
Happy to help.

Comment (raisedadead) [reply .github/workflows/e2e-playwright.yml:79]: Perfect, that's what I was looking for. Thanks for clarifying.

> If, I allow parallel runs, the CLI runs multiple browsers at the same time against the same localhost server running on port 8000.

This is an expected behavior and will obviously be flaky, and we do not what to test for this scenario at all. It is not real-world too. The "same user" is not executing the scenario in parallel on the same server-client config. Correct?

Can you add the strategy for parallel runs for three browsers? The runs should be for one browser and server in a 1:1 config within each job.

Here are the user stories as a developer:

1. When running locally, I can pass a param that runs the tests for a browser or default to Chrome.
2. When running in CI, I can run separate jobs with their servers and one of the three browsers as clients. The spec needs to be run within the jobs for the chosen browser.
3. In both cases, the "choice" of the browser is done at the config level (user chooses, or CI decides) and not automatically by Playwright.

Comment (raisedadead) [reply .github/workflows/e2e-playwright.yml:79]: To be clear the parallelism is offloaded to a higher layer and need not be coupled with playwright at all.

@hassaanp can help you understand the logic. If not I am happy to explain in our next meeting.

Thanks again.

Comment (Ahmadkashif) [reply .github/workflows/e2e-playwright.yml:79]: @raisedadead 

I understand you perfectly now. 

I initially planned to do it but refrained, thinking we might need to be bare-bone simple for now.

my next question is, do we merge this much progress and continue with the action in a fresh PR?

Comment (Ahmadkashif) [reply .github/workflows/e2e-playwright.yml:79]: @raisedadead 

I have incorporated the changes here: https://github.com/freeCodeCamp/freeCodeCamp/pull/51220
